psm/build.rs

```diff
@@ -37,24 +61,58 @@ fn main() {
     let env = ::std::env::var("CARGO_CFG_TARGET_ENV").unwrap();
     let os = ::std::env::var("CARGO_CFG_TARGET_OS").unwrap();
     let endian = ::std::env::var("CARGO_CFG_TARGET_ENDIAN").unwrap();
-    let asm = if let Some((asm, canswitch)) = find_assembly(&arch, &endian, &os, &env) {
-        println!("cargo:rustc-cfg=asm");
-        if canswitch {
-            println!("cargo:rustc-cfg=switchable_stack")
-        }
-        asm
-    } else {
-        println!(
-            "cargo:warning=Target {}-{}-{} has no assembly files!",
-            arch, os, env
-        );
-        return;
-    };
+    // Handle cross compilation scenarios where we're using eg clang-cl
+    // from a non-windows host, as by default cc will automatically try and
+    // run the appropriate Microsoft assembler for the target architecture
+    // if we give it a .asm file
+    let is_windows_host = std::env::var("HOST")
+        .unwrap_or_default()
+        .contains("-windows-");
+
+    let asm =
+        if let Some((asm, canswitch)) = find_assembly(&arch, &endian, &os, &env, is_windows_host) {
+            println!("cargo:rustc-cfg=asm");
+            if canswitch {
+                println!("cargo:rustc-cfg=switchable_stack")
+            }
+            asm
+        } else {
+            println!(
+                "cargo:warning=Target {}-{}-{} has no assembly files!",
+                arch, os, env
+            );
+            return;
+        };
 
     let mut cfg = cc::Build::new();
     let msvc = cfg.get_compiler().is_like_msvc();
 
-    if !msvc {
+    if !msvc || !is_windows_host {
```

Comment:
Do we need to check either of these conditions as opposed to just verifying the compiler being used is `clang-cl`?

Reply:
No, since it essentially amounts to the same thing, I'll streamline that check.